### DISCUSSION TITLE
DRC and LVS steps

### DISCUSSION THREAD

**Discussion by stevvik** [The-OpenROAD-Project/OpenROAD-flow-scripts#571] [Q&A]
Hello!
I have found in documentation? that in ORFS Klayout is used for DRC and LVS steps and decks are available for open PDKs.
In flow Makefile I see drc: and lvs: steps, but they don't run by default and KLAYOUT_DRC_FILE and KLAYOUT_LVS_FILE are not defined.

My questions are:
1) How can I run DRC and LVS steps in ORFS?
2) Where can I get an example of DRC/LVS decks for open PDKs (like sky130) for KLayout?

**Comment by maliberty** [comment]
This is an incomplete experiment in setting that up for sky130 in the style of OpenLane.  For now I would suggest OL for those functions or if  you are inclined to contribute you could bring over their setup to ORFS.

**Reply by rovinski** [reply]
DRC and LVS should still work on nangate45 with `make drc` and `make lvs`, respectively. If you want to add support for DRC on a new platform `$(PLATFORM)`:

1. Ensure that a KLayout tech file (`.lyt`) exists for the platform. If not, create it.
2. Create (or copy) a KLayout `.lydrc` file at `flow/platforms/$(PLATFORM)/drc/$(PLATFORM).lydrc`
3. Add the following line to `flow/platforms/$(PLATFORM)/config.mk`: `export KLAYOUT_DRC_FILE = $(PLATFORM_DIR)/drc/$(PLATFORM).lydrc`
 You should then be able to run `make drc`

To add support for LVS:
1. Ensure that a KLayout tech file (`.lyt`) exists for the platform. If not, create it.
2. Create (or copy) a KLayout `.lylvs` file at `flow/platforms/$(PLATFORM)/lvs/$(PLATFORM).lylvs`
3. Add the following line to `flow/platforms/$(PLATFORM)/config.mk`: `export KLAYOUT_LVS_FILE = $(PLATFORM_DIR)/lvs/$(PLATFORM).lylvs`
4. Ensure that a CDL file exists for your cell library at `flow/platforms/$(PLATFORM)/cdl/$(PLATFORM).cdl`. If not, create it.
5. Add the following line to `flow/platforms/$(PLATFORM)/config.mk`: `export CDL_FILE = $(PLATFORM_DIR)/cdl/$(PLATFORM).cdl`

 You should then be able to run `make lvs`

**Reply by stevvik** [reply]
@rovinski , Thank you for the info!
I found KLayout DRC deck for nangate45 platform in ORFS and for sky130A in OpenLane, but I didn't find any LVS deck examle for Klayout. I only see LVS setup for NetGen in OpenLane.
Can you show me the example of KLayout LVS deck for any platform?

**Reply by rovinski** [reply]
Oh right, I guess the LVS deck did not get uploaded because I couldn't get it to work (probably some combination of settings that need to be set for proper extraction).

@laurentc2 has created several KLayout decks for open PDKs.

* https://github.com/laurentc2/FreePDK45_for_KLayout
* https://github.com/laurentc2/SKY130_for_KLayout
* https://github.com/laurentc2/ASAP7_for_KLayout

You may need to tweak some of the settings in order to get LVS to match. On the other hand, my last trial was more than a year ago so maybe a new version of KLayout will work better.
